salt/engines/stalekey.py

```diff
@@ -39,41 +39,103 @@
 
 def __virtual__():
     if not __opts__.get('minion_data_cache'):
-        return (False, 'stalekey engine requires minion_data_cache to be enabled')
+        return (False, 'stalekey engine requires '
+                       'minion_data_cache to be enabled')
     return True
 
 
 def _get_keys():
+    '''
+    Get the keys
+    '''
     keys = salt.key.get_key(__opts__)
     minions = keys.all_keys()
     return minions['minions']
 
 
+def _delete_keys(stale_keys, minions):
+    '''
+    Delete the keys
+    '''
+    wheel = salt.wheel.WheelClient(__opts__)
+    for k in stale_keys:
+        log.info('Removing stale key for %s', k)
+        wheel.cmd('key.delete', [salt.utils.stringutils.to_unicode(k)])
+        del minions[k]
+    return minions
+
+
+def _read_presence(presence_file):
+    '''
+    Read minion data from presence file
+    '''
+    error = False
+    minions = {}
+    if os.path.exists(presence_file):
+        try:
+            with salt.utils.files.fopen(presence_file, 'rb') as f:
+                minions = msgpack.load(f)
+
+                # ensure all keys are unicode, not bytes.
+                _minions = {}
+                for minion in minions:
+                    _minion = salt.utils.stringutils.to_unicode(minion)
+                    _minions[_minion] = minions[minion]
+
+        except IOError as e:
+            error = True
+            log.error('Could not open presence file %s: %s', presence_file, e)
+
+    return error, _minions
+
+
+def _write_presence(presence_file, minions):
+    '''
+    Write minion data to presence file
+    '''
+    error = False
+    try:
+        with salt.utils.files.fopen(presence_file, 'wb') as f:
+            msgpack.dump(minions, f)
+    except IOError as e:
+        error = True
+        log.error('Could not write to presence file %s: %s', presence_file, e)
+    return error
+
+
+def _running():
+    '''
+    Indicate if engine should continue running
+    Useful when running tests and we want to ensure
+    the engine only makes one pass
+    '''
+    return True
+
+
 def start(interval=3600, expire=604800):
+    '''
```

Comment:
I'm going to pull a @dwoz on you :)
For reference, https://github.com/saltstack/salt/pull/55047#discussion_r336696340

Let's make start accept `_check_running_callable=None`

Then the `while` code Bellow becomes:
```python
while True:
    # do stuff
    If callable(_check_running_callable) and _check_running_callable() is False:
        break
```

No?

Reply:
Not against the idea, just curious, what is the advantage?

Reply:
Actually the `_running` function was being used when I was attempting to run `start` from tests, which didn't work out.  So I can remove that.

Reply:
Not having a function for us just to patch during tests, ie, dependency injection.